File: backend/F1Fantasy/Controllers/StandingsController.cs
using F1Fantasy.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;

namespace F1Fantasy.Controllers;

[Authorize]
[ApiController]
[Route("api/[controller]")]
public class StandingsController : ControllerBase
{
    private readonly StandingsService _standingsService;
    private readonly ScoringService _scoringService;

    public StandingsController(
        StandingsService standingsService,
        ScoringService scoringService)
    {
        _standingsService = standingsService;
        _scoringService = scoringService;
    }

    [HttpGet("groups/{groupId}")]
    public async Task<IActionResult> GetStandings(int groupId, [FromQuery] string season = "2026")
    {
        try
        {
            var standings = await _standingsService.GetStandingsWithAutoRecalcAsync(groupId, season);
            return Ok(standings);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpPost("groups/{groupId}/recalculate")]
    public async Task<IActionResult> RecalculateStandings(int groupId, [FromQuery] string season = "2026")
    {
        try
        {
            await _standingsService.RecalculateStandingsAsync(groupId, season);
            return Ok(new { message = "Standings recalculated successfully" });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/detailed")]
    public async Task<IActionResult> GetDetailedStandings(int groupId, [FromQuery] string season = "2026")
    {
        try
        {
            var standings = await _standingsService.GetDetailedStandingsAsync(groupId, season);
            return Ok(standings);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/user/{userId}/breakdown")]
    public async Task<IActionResult> GetUserBreakdown(int groupId, string userId, [FromQuery] string season = "2026")
    {
        try
        {
            var breakdown = await _scoringService.CalculateDetailedScoresAsync(groupId, userId, season);
            return Ok(breakdown);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/me/breakdown")]
    public async Task<IActionResult> GetMyBreakdown(int groupId, [FromQuery] string season = "2026")
    {
        try
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "User ID not found in token" });
            }

            var breakdown = await _scoringService.CalculateDetailedScoresAsync(groupId, userId, season);
            return Ok(breakdown);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/controller/StandingsController.java
package no.f1fantasy.controller;

import java.security.Principal;
import java.util.List;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import no.f1fantasy.dto.ErrorResponse;
import no.f1fantasy.dto.SuccessResponse;
import no.f1fantasy.entity.Standing;
import no.f1fantasy.service.ScoringService;
import no.f1fantasy.service.StandingsService;

@RestController
@RequestMapping("/api/standings")
@PreAuthorize("isAuthenticated()")
@SuppressWarnings("null")
public class StandingsController {
    
    private final StandingsService standingsService;
    private final ScoringService scoringService;

    public StandingsController(StandingsService standingsService, ScoringService scoringService) {
        this.standingsService = standingsService;
        this.scoringService = scoringService;
    }

    private String getUserId(Principal principal) {
        if (principal == null) {
            throw new IllegalArgumentException("User ID not found");
        }
        return principal.getName();
    }

    @GetMapping("/groups/{groupId}")
    public ResponseEntity<?> getStandings(
            @PathVariable int groupId,
            @RequestParam(defaultValue = "2026") String season) {
        try {
            List<Standing> standings = standingsService.getStandingsWithAutoRecalc(groupId, season);
            return ResponseEntity.ok(standings);
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @PostMapping("/groups/{groupId}/recalculate")
    public ResponseEntity<?> recalculateStandings(
            @PathVariable int groupId,
            @RequestParam(defaultValue = "2026") String season) {
        try {
            standingsService.recalculateStandings(groupId, season);
            return ResponseEntity.ok(new SuccessResponse("Standings recalculated successfully"));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/detailed")
    public ResponseEntity<?> getDetailedStandings(
            @PathVariable int groupId,
            @RequestParam(defaultValue = "2026") String season) {
        try {
            List<Standing> standings = standingsService.getStandingsWithAutoRecalc(groupId, season);
            return ResponseEntity.ok(standings);
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/user/{userId}/breakdown")
    public ResponseEntity<?> getUserBreakdown(
            @PathVariable int groupId,
            @PathVariable String userId,
            @RequestParam(defaultValue = "2026") String season) {
        try {
            Map<String, Integer> breakdown = scoringService.calculateAllCategoryScores(groupId, userId, season);
            return ResponseEntity.ok(breakdown);
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/me/breakdown")
    public ResponseEntity<?> getMyBreakdown(
            @PathVariable int groupId,
            @RequestParam(defaultValue = "2026") String season,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            Map<String, Integer> breakdown = scoringService.calculateAllCategoryScores(groupId, userId, season);
            return ResponseEntity.ok(breakdown);
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }
}
